File: contestJava/contest/src/Controller/AbstractController.java
package Controller;

import Repository.IRepository;
import Validator.ControllerException;
import Validator.RepositoryException;
import Validator.ValidatorException;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.function.Predicate;

/**
 * Created by Sebi on 25-Nov-16.
 */
public abstract class AbstractController<E, ID> {
    protected IRepository<E, ID> repository;

    public AbstractController(IRepository repository){
        this.repository = repository;
    }

    /*
    Saves an entity
     */
    public void save(String... args) throws ControllerException, RepositoryException, ValidatorException {
        E element = formatElement(args);
        repository.save(element);
    }

    /*
    Removes an entity
     */
    public void delete(String id) throws RepositoryException, ControllerException {
        ID convertedId = formatId(id);
        repository.delete(convertedId);
    }

    /*
    Updates an entity
     */
    public void update(String id, String... newElement) throws RepositoryException, ValidatorException, ControllerException{
        ID convertedId = formatId(id);
        E convertedElement = formatElement(newElement);
        repository.update(convertedId, convertedElement);
    }

    /*
    Returns the number of elements in repository
     */
    public int size(){
        return repository.size();
    }

    /*
    Returns the entity with the given id
     */
    public E getById(String id) throws ControllerException {
        ID convertedId = formatId(id);
        return repository.getById(convertedId);
    }

    /*
    Deletes all entities
     */
    public void clearAll(){
        repository.clearAll();
    }

    /*
    Returns all entities
     */
    public List<E> getAll(){
        return repository.getAll();
    }

    /*
    Generic sort, using a comparator
    */
    public <E> List<E> sortGeneric(List<E> list, Comparator<E> comparator){
        List<E> result = new ArrayList<E>();
        list.stream().sorted(comparator).forEach(result::add);
        return result;
    }

    /*
    Generic filter, using a predicate
     */
    public <E> List<E> filterGeneric(List<E> list, Predicate<E> predicate){
        List<E> result = new ArrayList<E>();
        list.forEach(x -> { if (predicate.test(x)) result.add(x);});
        return result;
    }

    /*
    Formats the element - will be implemented by each controller
     */
    public abstract E formatElement(String... args) throws ControllerException;

    /*
    Formats the id - will be implemented by each controller
     */
    public abstract ID formatId(String id) throws ControllerException;
}

File: contestC#/contest/contest/Controller/AbstractController.cs
using contest.Domain;
using contest.Repository;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace contest.Controller
{
    /*
     * Abstract Controller class 
     */
    public abstract class AbstractController<E, ID> where E : HasID<ID>
                                                    where ID : IComparable<ID>
    {
        //reference to repository
        protected IRepository<E, ID> repository;

        /*
         * Constructor
         */
        public AbstractController(IRepository<E, ID> repository)
        {
            this.repository = repository;
        }

        /*
         * Adds an entity
         */
         public void add(params string[] fields)
        {
            E entity = formatElement(fields);
            repository.add(entity);
        }

        /*
         * Removes an entity
         */
         public void delete(string id)
        {
            ID idEnt = formatId(id);
            repository.delete(idEnt);
        }

        /*
         * Updates an entity 
         */
         public void update(string id, params string[] fields)
        {
            ID idEnt = formatId(id);
            E entity = formatElement(fields);
            repository.update(idEnt, entity);
        }

        /*
         * Returns the entity with the given id 
         * or null if it is not found
         */
         public E getById(string id)
        {
            ID idEnt = formatId(id);
            return repository.getById(idEnt);
        }

        /*
         * Returns all entities 
         */
         public List<E> getAll()
        {
            return repository.getAll();
        }

        /*
         * Generic filter method
         */
         public List<E> genericFilter(List<E> entities, Predicate<E> filter)
        {
            return entities.FindAll(filter);
        }

        /*
         * formats the element 
         */
        public abstract E formatElement(params string[] fields);

        /*
         * formats the id 
         */ 
        public abstract ID formatId(string id);
    }
}
